### DISCUSSION TITLE
请教一个关于Task的问题

### DISCUSSION THREAD

**Discussion by osience** [BYJRK/DotNet-Discussions#179] [C# 技术交流]
我想实现一个异步往本地文本文件写日志的功能，代码如下

```c#
static Queue<string> logQueue = new Queue<string>();

static Task runTask = null;

static async Task WriteLogAsync(string logMessage)
{
    try
    {
        logQueue.Enqueue(logMessage);
        if (runTask == null)
        {
            runTask = Task.Run(() =>
            {
                while (logQueue.Count > 0)
                {
                    string strLog = logQueue.Dequeue();
                    WriteLogToFile(strLog);
                }
            });
            await runTask;
            runTask.Dispose();
            runTask = null;
            if (logQueue.Count > 0)
            {
                WriteLogAsync(logMessage);
            }
        }
    }
    catch (Exception ex)
    {
        Console.WriteLine(ex);
    }
}
```

因为第一次用Task 不确定这样使用有没有问题，我自己测试打印好像没有问题，请大佬帮忙review一下，谢谢

**Comment by BYJRK** [comment]
你的代码问题比较严重。

1. 你在异步方法中使用一个线程不安全的 `Queue`，这是不合适的。应该考虑使用 `ConcurrentQueue` 等
2. 你对于 `runTask` 的使用存在很多误区，比如你在类内保存它的字段，这看起来并没有什么意义，因为在这个函数外似乎并没有什么地方会去使用它；不仅如此，对于一个 `Task`，使用 `Dispose()` 是不必要的
3. 你在该方法中递归调用自己，但是却采用了“一发即忘”的方式，这是非常不恰当的。你将完全无法知道异步任务的各种情况

抛开这些问题不说，首先你试图异步打日志这一点就很奇怪。除非打日志相当耗时，否则你完全可以考虑直接同步调用日志方法。即便你真的有必要不在当前线程上做这件事情，那么你也应该考虑另开一个线程，专门用来打日志。这里可以考虑使用 `BlockingCollection`，因为它自带信号量，而不需要你去轮询或单独维护信号量，也不需要考虑线程安全的问题。

另外，你似乎自己写了一个写日志的方法，这其实也是不妥的。因为现在有很多相当成熟的日志工具，比如 Serilog、NLog、Log4net 等。你不应该考虑在这种事情上自己造轮子。

**Reply by osience** [reply]
这个 runTask 我的想法是，如果有一个 Task 正在写 log 的话，就不需要再启动一个新 Task 了，往队列尾部追加就行了。看来是我对 Task 的使用场景理解有问题。

另外我也感觉使用 Queue 会有线程安全问题，但是不知道是不是我测试方法有问题，我在 Task 写日志的时候，同时去往队列追加新日志竟然是没有问题的，这也是我困惑的地方😵‍💫
感谢您耐心的回复，第一次写异步相关的东西真的有点无从下手哈哈
